pkg/services/datasources/service/datasource.go

```diff
@@ -252,6 +295,89 @@ func (s *Service) AddDataSource(ctx context.Context, cmd *datasources.AddDataSou
 	})
 }
 
+// This will valid validate the instance settings and mutate the cmd with the processed values
+func (s *Service) mutate(ctx context.Context, pluginContext backend.PluginContext, settings *backend.DataSourceInstanceSettings) (*backend.DataSourceInstanceSettings, error) {
+	operation := backend.AdmissionRequestCreate
+
+	if len(settings.Name) > maxDatasourceNameLen {
+		return nil, datasources.ErrDataSourceNameInvalid.Errorf("max length is %d", maxDatasourceNameLen)
+	}
+
+	if len(settings.URL) > maxDatasourceUrlLen {
+		return nil, datasources.ErrDataSourceURLInvalid.Errorf("max length is %d", maxDatasourceUrlLen)
+	}
```

Comment:
isn't this validation, not mutation?

Reply:
I still think it's more clear if the mutation and validation are as isolated as possible, in this case, I don't see why we wouldn't want to keep doing `validateFields` as we have right now.

Reply:
the global validateFields is good for things that exist across everything -- max length on URL and name ✔️  but we need plugin hooks to know that the URL should not even exist!

Reply:
yes, I am not saying to remove the hooks, I am just saying to restore `validateFields` (with the check on max length) out of the `mutate` functionality.

Reply:
I renamed the function "prepareInstanceSettings" -- maybe that helps clarify that it is a combination of global validation and a mutation hook (eventually whatever is configured)